### PR TITLE
fix handling same colName for existing colName, alias colName and group by colName 

### PR DESCRIPTION
**Pull request by jennifersp**
Fixed a case of table has `col1` and it's used in the query for referencing this column, and using it for alias name for this column and using it for groupBy column.
Example is `SELECT col1, col1 AS col1 FROM tab1 GROUP BY col1`
Follows MySql `ONLY_FULL_GROUP_BY` mode

### CODE REVIEW

File: enginetest/script_queries.go [hunk 1]
```diff
@@ -602,6 +602,29 @@ var ScriptTests = []ScriptTest{
...
+			{
+				Query:    "SELECT col1, col1 AS col1 FROM tab1 GROUP BY col1",
+				Expected: []sql.Row{{6, 6}, {57, 57}, {44, 44}},
+			},
+			{
+				Query:    "SELECT - col2, - col2 * - + col2 + col2 AS col2 FROM tab1 GROUP BY col2",
+				Expected: []sql.Row{{-8, 72}, {-45, 2070}, {-71, 5112}},
+			},
+			{
+				Query:    "SELECT 9 * col2 + col2 + 30 * - col0 col1, col0 + - 1 * + 0 * - ( - col0 ) AS col0 FROM tab1 cor0 GROUP BY col0, col2",
```
Comment (max-hoffman) [inline enginetest/script_queries.go:623]: some ideas for potential edge cases:
```SQL
mysql> select col1 as col3 from tab1 group by col3;
+------+
| col3 |
+------+
|    6 |
|   57 |
|   44 |
+------+
mysql> select (1) as col1, b.col1 as col1 from tab1 as a join tab1 as b where a.col1 = b.col1 group by a.col1;
+------+------+
| col1 | col1 |
+------+------+
|    1 |    6 |
|    1 |   57 |
|    1 |   44 |
+------+------+
mysql> SELECT (SELECT col1 as col2) as col1, col1 AS col1 FROM tab1 GROUP BY col1;
+------+------+
| col1 | col1 |
+------+------+
|    6 |    6 |
|   57 |   57 |
|   44 |   44 |
+------+------+
mysql>WITH cte1(col1) as (select col2 from tab1) select cte1.col1 from cte1 GROUP BY col2;
+------+
| col1 |
+------+
|    8 |
|   45 |
|   71 |
+------+
mysql> WITH cte1(col1) as (select col2 from tab1) select cte1.col1 as col1, tab1.col1 AS col1 FROM tab1 join cte1 where cte1.col1 = tab1.col2 GROUP BY tab1.col1, cte1.col1;
+------+------+
| col1 | col1 |
+------+------+
|    8 |    6 |
|   45 |   57 |
|   71 |   44 |
+------+------+
mysql> select col1 from tab1 union select col1 from tab1 group by col1;
+------+
| col1 |
+------+
|    6 |
|   57 |
|   44 |
+------+
mysql> select (1) as col1, col1 from tab1 union select col2, col1 from tab1 group by col1, col2;
+------+------+
| col1 | col1 |
+------+------+
|    1 |    6 |
|    1 |   57 |
|    1 |   44 |
|    8 |    6 |
|   45 |   57 |
|   71 |   44 |
+------+------+
```

File: sql/analyzer/validation_rules.go [hunk 2]
```diff
@@ -246,10 +246,29 @@ func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (s
 			groupBys = append(groupBys, expr.String())
 		}
 
+		var curAliases = make(map[string]sql.Expression)
+		if p, ok := n.Child.(*plan.Project); ok {
```
Comment (max-hoffman) [inline sql/analyzer/validation_rules.go:250]: there are a variety of node types that could be a `*plan.GroupBy` child. An order by might insert a sort node, for example:
```
tmp3> explain select * from b group by x order by x;
+---------------------------------------+
| plan                                  |
+---------------------------------------+
| GroupBy                               |
|  ├─ SelectedExprs(b.x)                |
|  ├─ Grouping(b.x)                     |
|  └─ Sort(b.x ASC)                     |
|      └─ Projected table access on [x] |
|          └─ Exchange(parallelism=12)  |
|              └─ Table(b)              |
+---------------------------------------+
```

Comment (max-hoffman) [reply sql/analyzer/validation_rules.go:250]: In the context of validating group by functional dependencies, this means 1) a node TraverseUp might be necessary to resolve all of the child nodes, and 2) an expression Traverse up might be necessary to find the expression dependencies for each child. This [copy function](https://github.com/dolthub/go-mysql-server/blob/30a0c6358b4500db684b0d3fb478521c9a057d66/sql/analyzer/analyzer.go#L507) (not merged yet) shows one way to do this sort of traversal.

File: sql/analyzer/validation_rules.go [hunk 3]
```diff
@@ -260,6 +279,26 @@ func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (s
 	return n, nil
 }
 
+func expressionReferencesAlias(expr sql.Expression) map[string]sql.Expression {
+	var children = make(map[string]sql.Expression)
+	sql.Inspect(expr, func(expr sql.Expression) bool {
```
Comment (max-hoffman) [inline sql/analyzer/validation_rules.go:284]: i might be mistaken, but does the `false` short-circuit the walk and potentially miss multiple child expressions?  For example, this should fail, but if we are ending the walk after finding the first `GetField` we may be mistakenly OK'ing the query:
```SQL
mysql> select col1 + col2 from tab1 group by col1;
```

**Review by max-hoffman — COMMENTED** [review]
Good start! Alias resolution has a lot of edge cases and I tried to list as many as I could, but that doesn't necessarily mean you have to fix all of these before merging. Tests for which ones we don't support are also helpful, I defer to Zach for scoping how deep to dig.

File: sql/analyzer/validation_rules.go (hunk 2 above)
Comment (zachmu) [reply sql/analyzer/validation_rules.go:250]: Any way you look at this, the above plan is a bug. `ORDER BY` always applies to a the final result set, it cannot be the child of something like a grouping operation. Can you file a bug for that @max-hoffman ?

Comment (max-hoffman) [reply sql/analyzer/validation_rules.go:250]: If we had a grouping operator that took advantage of sorted input it would make sense, but otherwise yeah.

File: sql/analyzer/validation_rules.go [hunk 4]
```diff
@@ -246,10 +246,29 @@ func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (s
...
+		}
+
 		for _, expr := range n.SelectedExprs {
 			if _, ok := expr.(sql.Aggregation); !ok {
 				if !expressionReferencesOnlyGroupBys(groupBys, expr) {
-					return nil, ErrValidationGroupBy.New(expr.String())
+					depExprs := expressionReferencesAlias(expr)
+					for _, dep := range depExprs {
+						if !expressionReferencesOnlyGroupBys(groupBys, dep) {
+							if e, ok := curAliases[dep.String()]; ok {
```
Comment (zachmu) [inline sql/analyzer/validation_rules.go:264]: Looks like you are using the string value, just computing it again. Use the 2-value range expression instead. Or, change the `expressionReferencesAlias` method to return just a slice of sql.Expr

File: sql/analyzer/validation_rules.go [hunk 5]
```diff
@@ -270,6 +309,25 @@ func expressionReferencesOnlyGroupBys(groupBys []string, expr sql.Expression) bo
 			if stringContains(groupBys, expr.String()) {
 				return false
 			}
+			return true
+		case *expression.GetField:
+			for _, groupBy := range groupBys {
```
Comment (zachmu) [inline sql/analyzer/validation_rules.go:314]: You can call stringContains for both these checks, rather than looping over groupBys

File: sql/analyzer/validation_rules.go [hunk 6]
```diff
@@ -270,6 +309,25 @@ func expressionReferencesOnlyGroupBys(groupBys []string, expr sql.Expression) bo
...
+						return true
+					}
+				} else {
+					if groupBy == expr.Name() {
+						return true
+					}
+				}
+			}
+
+			if len(expr.Children()) == 0 {
```
Comment (zachmu) [inline sql/analyzer/validation_rules.go:326]: Again no reason for this check

**Review by zachmu — COMMENTED** [review]
I think this is fine, a few comments:

* Write the test cases Max suggested
* Look through `errorQueries` for queries that should be fixed according to these rules
